### PR TITLE
support for proxy contracts for `from_etherscan`

### PR DESCRIPTION
**Pull request by mo-anon**
Please beware with me, im new to all this.

### What I did
- Modified `fetch_abi_from_etherscan` to use the ABI of the implementation contract if it is a proxy, enabling the reading of proxy contracts.
- Added a `from_etherscan_abi` function which returns the actual ABI. I feel like this might be useful (will remove if not needed).

### How I did It
Used the [Etherscan API](https://docs.etherscan.io/api-endpoints/contracts#get-contract-source-code-for-verified-contract-source-codes) to check if the contract is a proxy. If yes, the ABI of the implementation contract is used.
_NOTE_: Im very positive there are more efficient ways to resolve this as i call the api twice, but it's the only solution I can think of at the moment lol.

### How to verify It
Added some simple tests. However, testing/using `from_etherscan` is a pain because the ABI parser raises an error when the ABI _apparently_ has an invalid length, e.g., `uint256[]` or `address[]`. As i see, this is getting fixed (https://github.com/vyperlang/titanoboa/pull/115)

### Description for the changelog
- Added proxy contract support for `from_etherscan`.
- Added `from_etherscan_abi`, which retrurns the full ABI of a contract.

### Cute Animal Picture

Insert a link to a cute animal picture here

### CODE REVIEW

File: boa/explorer.py [hunk 1]
```diff
@@ -12,13 +12,30 @@ def fetch_abi_from_etherscan(
 ):
     address = Address(address)
 
-    params = dict(module="contract", action="getabi", address=address)
+    params = dict(module="contract", action="getsourcecode", address=address)
```
Comment (DanielSchiavini) [inline boa/explorer.py:15]: If this is only called to get the address, let's create a function `_get_implementation_address` so the purpose can be more clear

File: boa/explorer.py [hunk 2]
```diff
@@ -12,13 +12,30 @@ def fetch_abi_from_etherscan(
...
     if api_key is not None:
         params["apikey"] = api_key
 
     res = SESSION.get(uri, params=params)
     res.raise_for_status()
+    source_data = res.json()
+
+    if int(source_data["status"]) != 1:
+        raise ValueError(f"Failed to retrieve data from API: {data}")
```
Comment (DanielSchiavini) [inline boa/explorer.py:24]:
```suggestion
        raise ValueError(f"Failed to retrieve source from Etherscan: {data}")
```

File: tests/unitary/test_from_etherscan.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+crvusd = '0xf939E0A03FB07F59A73314E73794Be0E57ac1b4E'
+voting_agent = "0xE478de485ad2fe566d49342Cbd03E49ed7DB3356"
+
+
+def test_from_etherscan_abi():
+
+    boa.env.fork(os.environ["ALCHEMY_MAINNET_ENDPOINT"])
+    abi = boa.from_etherscan_abi(crvusd, api_key=os.environ["ETHERSCAN_API_KEY"])
+
+    return abi
```
Comment (DanielSchiavini) [inline tests/unitary/test_from_etherscan.py:12]: it's not a real test if there's no assertion (same for all tests here). For example:
```suggestion
    assert abi.totalSupply() > 0
```

File: tests/unitary/test_from_etherscan.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import os
+
+import boa
+
+crvusd = "0xf939E0A03FB07F59A73314E73794Be0E57ac1b4E"
+voting_agent = "0xE478de485ad2fe566d49342Cbd03E49ed7DB3356"
+
+
+def test_from_etherscan():
+    boa.env.fork(os.environ["ALCHEMY_MAINNET_ENDPOINT"])
```
Comment (bout3fiddy) [inline tests/unitary/test_from_etherscan.py:10]: these tests should be in tests/forked:

https://github.com/vyperlang/titanoboa/blob/ddd3db55a0e4ee24f1386306d70e617197ad41ae/tests/integration/fork/conftest.py

and should be using the 'forked' fixture.

File: boa/interpret.py [hunk 5]
```diff
@@ -126,4 +126,11 @@ def from_etherscan(
     return ABIContractFactory.from_abi_dict(abi, name=name).at(addr)
 
 
+def from_etherscan_abi(
```
Comment (charles-cooper) [inline boa/interpret.py:129]: i'm not sure why this function is needed

**Comment by charles-cooper** [comment]
i guess we are going to need add an etherscan api key to the CI

File: boa/interpret.py (hunk 5 above)
Comment (charles-cooper) [reply boa/interpret.py]: i removed it in [82f6f7d](https://github.com/vyperlang/titanoboa/pull/122/commits/82f6f7d1a87f737f04ea1be1d4ac8c0342fa0c11). we can always add back alter

File: boa/explorer.py [hunk 6]
```diff
@@ -10,18 +10,42 @@
...
     data = res.json()
 
     if int(data["status"]) != 1:
         raise ValueError(f"Failed to retrieve data from API: {data}")
 
     return json.loads(data["result"].strip())
+
+
+def _get_implementation_address(
+    address: str, uri: str = "https://api.etherscan.io/api", api_key: str = None
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply boa/explorer.py:24]: i removed in [f52bb3e](https://github.com/vyperlang/titanoboa/pull/122/commits/f52bb3e1f97a47e79103930ba701c55f9a5e3f62)

**Comment by charles-cooper** [comment]
ok i fixed all the little issues. merging, thanks!

**Pull request merged**